posthog/models/exchange_rate/sql.py

```diff
@@ -0,0 +1,166 @@
+import csv
+import datetime
+import os
+
+from posthog.clickhouse.cluster import ON_CLUSTER_CLAUSE
+from posthog.clickhouse.table_engines import (
+    MergeTreeEngine,
+    ReplicationScheme,
+)
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_PASSWORD
+
+
+# This loads historical data from `historical.csv`
+# and generates a dictionary containing all entries for all dates and all currencies
+# from January 1st 2000 to December 31st 2024
+#
+# All of the rates are in comparison to the USD at that time.
+# There's no inflation adjustment, that would be nonsense.
+# If you want to convert from currency A to currency B, you need to:
+# 1. Convert A to USD
+# 2. Convert USD to B
+#
+# This is easily achieved by: `amount` B = `amount` A * `rate_A` / `rate_B`
+#
+# This CSV was originally downloaded from https://github.com/xriss/freechange/blob/master/csv/usd_to_xxx_by_day.csv
+# and then slightly optimized:
+# 1. Remove all dates older than 2000-01-01
+# 2. Truncate all rates to 4 decimal places
+# 3. Remove USD because it's the base currency, therefore it's always 1:1
+# 4. Add UYU rate for 2000-01-01 because we only had from 2010 onwards, based on https://fred.stlouisfed.org/series/FXRATEUYA618NUPN
+#
+# The resulting CSV is stored in `historical.csv`
+#
+# This won't return values for dates where we didn't have a value.
+# When querying the table/dictionary, you'll need to look for the previous closest date with a value.
+def HISTORICAL_EXCHANGE_RATE_DICTIONARY():
+    rates_dict = {}
+    currencies = []
+
+    # Load the CSV file
+    with open(os.path.join(os.path.dirname(__file__), "historical.csv")) as f:
+        reader = csv.reader(f)
+
+        # Get header row with currency codes
+        currencies = next(reader)
+        currencies = [c.strip() for c in currencies]
+
+        # Parse each row
+        for row in reader:
+            if not row:  # Skip empty rows
+                continue
+
+            date_str = row[0].strip()
+            try:
+                # Parse the date
+                date = datetime.datetime.strptime(date_str, "%Y-%m-%d")
+                date_key = date.strftime("%Y-%m-%d")
+
+                # Create dictionary for this date
+                rates = {}
+                for i, value in enumerate(row[1:], 1):
+                    currency = currencies[i]
+
+                    # Only add non-empty values
+                    value = value.strip()
+                    if value:
+                        try:
+                            rates[currency] = float(value)
+                        except ValueError:
+                            # Just ignore non-numeric values
+                            pass
+
+                rates_dict[date_key] = rates
+            except ValueError:
+                # Skip rows with invalid dates
+                continue
+
+    # Fill in missing dates
+    start_date = datetime.datetime(2000, 1, 1)
```

Comment:
Just for a clarification, the exchange rates were published for what time zone? UTC or some US one?

Reply:
The idea here is to store it in UTC at all times. I'll double-check the file we downloaded is using UTC timezones. It's not the most accurate file, but I don't expect us to *need* extremely accurate historical data, rates won't change much between midnight UTC and midnight PST/EST.

Reply:
I can actually remove this `"Fill in missing dates"` because it'll be empty anyway, it was helpful for debugging, but now that this is working, we don't need it anymore

Reply:
AFAIK the rates usually are there only when official market is running, and official rates usually are what national banks are publishing the day after.